_posts/2023-08-23-developer-ergonomics-hier-kommen-angular-signals/2023-08-23-developer-ergonomics-hier-kommen-angular-signals.md

```diff
@@ -0,0 +1,196 @@
+
+Die API eines Writable Signals bietet vier Methoden:
+- `set` setzt einen neuen Wert des Signals
+- `update` aktualisiert den Wert des Signals basierend auf dem aktuellen Wert
+- `mutate` aktualisiert den Wert des Signals mit einer direkten Änderung
+- `asReadonly` gibt das Signal als Readonly-Signal zurück, die oben genannten Methoden stehen dann nicht zur Verfügung.
+
+Lass uns nun ein Writable Signal und damit unser erstes Signal erstellen.
+
+Schritt 1: Wir importieren die Funktion signal von `@angular/core`. Diese Funktion erlaubt es uns, ein Writable Signal zu erstellen. Füge dazu einfach den Import am Anfang der Datei hinzu, in der Du das Computed Signal erstellen möchtest:
+
+```
+import { signal } from "@angular/core"
+```
+
+Schritt 2: Nun können wir das Writable Signal erstellen. Dazu nutzen wir die importierte Funktion signal.
+
+Beim Erstellen eines Writable Signals lässt sich der initiale Wert des Signals mitgeben, ähnlich wie bei der Instanziierung von einem `BehaviourSubject`.
+
+```
+// Signals
+import { signal } from "@angular/core";
+
+invoices = signal([]);
+```
+
+Schauen wir uns das Beispiel oben an, sehen wir, dass die API von Signals sehr ähnlich der von RxJS ist. Ein erster Unterschied lässt sich erkennen im Zugriff auf den jeweiligen Wert. Während der Zugriff auf Observables / Subjects im Template z. B. über die async-Pipe funktioniert, lässt sich der Zugriff auf Signals über einen einfachen und direkten Funktionsaufruf lösen.
+
+Schritt 3: Wir greifen im Template unserer Komponente auf den Wert des Signals zu.
+
+```
+<!-- Signals -->
+<p>Anzahl Rechnungen: {{ invoices().length }}</p>
+```
+
+Du kannst erkennen, dass der Zugriff hier ähnlich, wenn auch anders, funktioniert. Es sei gesagt, dass der Zugriff auf ein BehaviourSubject auch anders funktionieren kann. Wir beschränken uns hier allerdings auf diese Variante.
+
+Nun lass uns dem Signal auch einen Wert geben.
+
+Schritt 4: Wir implementieren eine Methode zum Setzen des Werts des Signals.
+
+```
+addInvoice(invoice: Invoice) {
+    // Weg 1: Mutate
+    this.invoices.mutate(invoices => invoices.push(invoice));
+
+    // Weg 2: Update
+    this.invoices.update(invoices => [...invoices, invoice]);
+
+    // Weg 3: Set
+    this.invoices.set([...this.invoices(), invoice]);
+}
+```
+
+Alle drei Wege fügen die Rechnung zur Liste hinzu. Die Methoden `mutate` und `update` bieten sich an, wenn der neue Wert vom aktuellen Wert abhängt. `update` ermöglicht es, mit Immutables zu arbeiten (z. B. für Performance-Gründe).
```

Comment:
_Personal Note_

Aus meiner Sicht, kann man derzeit fast nur mit `update` arbeiten, obwohl ich `mutate` angenehmer finde für das Aktualisieren von Werten.

Der Grund ist, dass `mutate` noch nicht mit Kind-Komponenten funktioniert, die `OnPush`-ChangeDetection verwenden.
Das wird in Zukunft funktionieren, sobald die _Signal-Story_ komplett umgesetzt ist.